### ISSUE TITLE
Define end-2-end testing scenario for Mini

### ISSUE THREAD

**Issue by netrome** [stacks-network/sbtc#11]
From [testing dialogue](https://github.com/stacks-network/stacks-blockchain/discussions/3787#discussioncomment-6577287)

To prepare for e2e integration testing for Mini, we'd like to define a first scenario.  For example,

1. configure and run a bitcoin and a stacks node
2. configure and run signers on different machines
3. wait for this...
4. observe this...
5. shutdown the machines

Once this is defined we can collaborate with the testing group to set it up.

### Definition of done
An initial testing scenario for Mini is set up

**Comment by CAGS295** [comment]
Is [this](https://github.com/Trust-Machines/stacks-sbtc/tree/main/sbtc-mini) mini?

**Comment by netrome** [comment]
> Is [this](https://github.com/Trust-Machines/stacks-sbtc/tree/main/sbtc-mini) mini?

Yes. We have docs in progress. It's being rebranded as sBTC 0.1/Developer release.

**Comment by CAGS295** [comment]
**WIP**

1. Select a network target and network settings.
2. Start a global orchestrator using DevnetJS.
3. Deploy smart contract tests (if ephemeral network).
4. Await or walkthrough until preconditions are met. e.g. epochs?
5. Execute tests.
    - Non-consensus tests run in parallel or in lockstep.
    - Consensus tests run sequentially.
7. Destroy smart contract tests (if feasible).
8. Terminate the orchestrator.

**Comment by CAGS295** [comment]
@netrome can you point me to the current e2e/integration solution for sBTC?

**Comment by netrome** [comment]
> @netrome can you point me to the current e2e/integration solution for sBTC?

There isn't one. This is the first stab at it.

**Comment by bitfari** [comment]
Multiple issues missing, for example: we need to store the addresses of generated nodes, plus have configurable variables to allot different waiting times across envs (test, dev, main). Currently drafting something a bit more concrete.

**Comment by CAGS295** [comment]
> Multiple issues missing, for example: we need to store the addresses of generated nodes, plus have configurable variables to allot different waiting times across envs (test, dev, main). Currently drafting something a bit more concrete.

It sounds like I've got the scope of the tests wrong. Your input makes me think that you actually need to deploy a network on demand and make sure through tests that consensus works as expected. Put simply, do the integration tests involve a single node or many nodes?

**Comment by CAGS295** [comment]
@friedger Any thoughts, we were unknowingly working on this in parallel. I've got as far as implementing the bootstrap call as a deployment. I'm currently trying to figure out what kind of integration test requirements we would need.

**Comment by bitfari** [comment]
> > Multiple issues missing, for example: we need to store the addresses of generated nodes, plus have configurable variables to allot different waiting times across envs (test, dev, main). Currently drafting something a bit more concrete.
> 
> It sounds like I've got the scope of the tests wrong. Your input makes me think that you actually need to deploy a network on demand and make sure through tests that consensus works as expected. Put simply, do the integration tests involve a single node or many nodes?

No. Actually, the scope is spot on, just trying to make sure we don't miss any crucial checks or assertions. Also, we need to clarify, do we just want a general idea of test scenarios or do we want to list key test cases as well with all the inputs? 

For reference, see this one mentioned by @friedger https://github.com/Trust-Machines/stacks-sbtc/pull/655, multiple nodes plus a list of variables to check later. I think it would be valuable to include that level of detail.

**Comment by CAGS295** [comment]
> > > Multiple issues missing, for example: we need to store the addresses of generated nodes, plus have configurable variables to allot different waiting times across envs (test, dev, main). Currently drafting something a bit more concrete.
> > 
> > 
> > It sounds like I've got the scope of the tests wrong. Your input makes me think that you actually need to deploy a network on demand and make sure through tests that consensus works as expected. Put simply, do the integration tests involve a single node or many nodes?
> 
> No. Actually, the scope is spot on, just trying to make sure we don't miss any crucial checks or assertions. Also, we need to clarify, do we just want a general idea of test scenarios or do we want to list key test cases as well with all the inputs?
> 
> For reference, see this one mentioned by @friedger [Trust-Machines/stacks-sbtc#655](https://github.com/Trust-Machines/stacks-sbtc/pull/655), multiple nodes plus a list of variables to check later. I think it would be valuable to include that level of detail.

I think we want the former, and we could get there by extrapolating from an initial concrete case. I'll use friedger's tests for now.

**Comment by AshtonStephens** [comment]
@CAGS295 and @friedger - right now a number of us are working on issue https://github.com/stacks-network/sbtc/issues/53. We're trying to get a local devnet up and running for us to run integration tests against during the NY Hackathon. Are the integration tests you're referencing using a local devnet?

**Comment by friedger** [comment]
@AshtonStephens yes, local devnet. It is really easy

**Comment by CAGS295** [comment]
There are plenty of examples of integration tests using devnetJS instead of clarinet integrate.
[Here](https://github.com/hirosystems/stacks-e2e-testing) is another useful resource other than the ones friedger mentioned in #53.

**Comment by CAGS295** [comment]
@friedger can you take a quick look at the reference implementation? I think this aligns well with what you were working on related to integration and flow tests. What is your take? Does it look useful to you? Would a flow test of yours be a good first integration test to try? I'm aware some features may be missing but that would be part of the task.

@moodmosaic Could you also take a quick look at the source code? Do you think this could work well for property-based tests?

@AshtonStephens Not sure how the integration tests look on the rust side, but we could adopt this same approach in Rust. Mainly by overhauling DevnetJS. Is there a rust SDK similar to StacksJS?

**Comment by CAGS295** [comment]
@sergey-shandar I would like to wrap up this issue asap, it is starting to feel stale. Anything else I can add? I feel like we didn't quite meet the goal here 100%. We still don't have a meaningful testing scenario so I may just port one from Friedger's flow tests or pick one up from the other stacks' repos integration tests. Nevertheless, this opens up many actionables.

**Comment by friedger** [comment]
Furthermore, this is about mini, and we want sBTC 0.2

**Comment by CAGS295** [comment]
@netrome I will try to come up with a meaningful integration test. Do you know someone that can provide a description for the sbtc-mini user's happy path?

**Comment by CAGS295** [comment]
I just saw that sbtc-mini was discontinued. Should we open a new issue for sBTC 0.2?

**Comment by netrome** [comment]
I think we could repurpose this one. I'd like to have good integration testing for the new developer release as well. The happy path would be:

1. A user creates a deposit request https://stacks-network.github.io/sbtc-docs/sbtc-operations/bitcoin-transactions.html
2. The sBTC DR system mints the requested amount of sBTC to the users account.
3. The user creates a withdrawal request
4. The sBTC DR system burns the requested amount of sBTC from the users account.
5. Upon a successful burn, the system then fulfills the withdrawal request by transferring the requested amount of BTC back to the users account.

**Comment by CAGS295** [comment]
Can you point me to the source code? I already have a basic example of an automated script for integration tests. I'd like to draw more requirements from the source code.

**Comment by netrome** [comment]
This is the old source code: https://github.com/Trust-Machines/stacks-sbtc/tree/main/stacks-coordinator
But we're currently revamping it and porting the functionality here: https://github.com/stacks-network/sbtc/pull/72

**Comment by AshtonStephens** [comment]
I'm closing this issue because Mini is deprecated and https://github.com/stacks-network/sbtc/issues/95 is the replacement.
